src/persist-client/src/internal/gc.rs

```diff
@@ -201,277 +205,442 @@ where
+                continue;
             }
-            state_count += 1;
-
-            match state.seqno.cmp(&req.new_seqno_since) {
-                Ordering::Less => {
-                    state.collections.trace.map_batches(|b| {
-                        for part in b.parts.iter() {
-                            // It's okay (expected) if the key already exists in
-                            // deleteable_batch_blobs, it may have been present in
-                            // previous versions of state.
-                            deleteable_batch_blobs.insert(part.key.to_owned());
-                        }
-                    });
+
+            // By our invariant, `states` should always begin on a rollup.
+            assert!(
+                gc_rollups.contains_seqno(&states.state().seqno),
+                "rollups = {:?}, state seqno = {}",
+                gc_rollups,
+                states.state().seqno
+            );
+
+            Self::find_removable_blobs(
+                states,
+                truncate_lt,
+                &machine.applier.metrics.gc.steps,
+                timer,
+                &mut batch_parts_to_delete,
+                &mut rollups_to_delete,
+            );
+
+            // After finding removable blobs, our state should be exactly `truncate_lt`,
+            // to ensure we've seen all blob deletions in the diffs needed to reach
+            // this seqno.
+            //
+            // That we can always reach `truncate_lt` given the live diffs we fetched
+            // earlier is a little subtle:
+            // * Our GC request was generated after `seqno_since` was written.
+            // * If our initial seqno on this loop was < `truncate_lt`, then our read
+            //   to `fetch_all_live_states` must have seen live diffs through at least
+            //   `seqno_since`, because the diffs were not yet truncated.
+            // * `seqno_since` >= `truncate_lt`, therefore we must have enough live
+            //   diffs to reach `truncate_lt`.
+            assert_eq!(states.state().seqno, truncate_lt);
+            // `truncate_lt` _is_ the seqno of a rollup, but let's very explicitly
+            // assert that we're about to truncate everything less than a rollup
+            // to maintain our invariant.
+            assert!(
+                gc_rollups.contains_seqno(&states.state().seqno),
+                "rollups = {:?}, state seqno = {}",
+                gc_rollups,
+                states.state().seqno
+            );
+
+            // Extra paranoia: verify that none of the blobs we're about to delete
+            // are in our current state (we should only be truncating blobs from
+            // before this state!)
+            states.state().map_blobs(|blob| match blob {
+                HollowBlobRef::Batch(batch) => {
+                    for live_part in &batch.parts {
+                        assert!(batch_parts_to_delete.get(&live_part.key).is_none());
```

Comment:
nit: if you `assert_eq!(... ,None)` instead, then we'd get the actual key on failure